File: Python/Student_Result_Processing_System/services/class_stats.py
from services.sorting import sort_students_by_total_marks

def get_class_topper(students):
    if len(students) == 0:
        return None
    ranked = sort_students_by_total_marks(students)
    return ranked[0]

def get_class_average(students):
    if len(students) == 0:
        return 0
    total = 0
    for student in students:
        total += student.get_average()
    return total / len(students)

def get_pass_percentage(students):
    if len(students) == 0:
        return 0
    pass_count = 0
    for student in students:
        if student.is_pass():
            pass_count += 1
    return (pass_count / len(students)) * 100

def print_class_summary(students):
    topper = get_class_topper(students)
    average = get_class_average(students)
    pass_pct = get_pass_percentage(students)

    print("--- Class Summary ---")
    if topper is None:
        print("No students to summarize.")
        return

    print(f"Class Topper: {topper._name} (Roll No: {topper._roll_no}, Total Marks: {topper.get_total_marks()})")
    print(f"Class Average: {average:.2f}")
    print(f"Pass Percentage: {pass_pct:.2f}%")

File: Student_Result_Processing_System/Services/ClassStats.cs
using Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Services
{
    public static class ClassStats
    {
        public static Student GetClassTopper(List<Student> students)
        {
            if(students.Count == 0)
            {
                return null;
            }
            List<Student> ranked = SortHelper.SortStudentsByTotalMarks(students);
            return ranked[0];
        }

        public static double GetClassAverage(List<Student> students)
        {
            if (students.Count == 0)
            {
                return 0;
            }

            double total = 0;
            foreach(Student s in students)
            {
                total += s.GetAverage();
            }
            return total / students.Count;
        }

        public static double GetPassPercentage(List<Student> students)
        {
            if (students.Count == 0)
            {
                return 0;
            }

            int passCount = 0;
            foreach(Student s in students)
            {
                if (s.isPass())
                {
                    passCount++;
                }
            }

            return ((double)passCount / students.Count) * 100;
        }

        public static void PrintSummary(List<Student> students)
        {
            Student topper = GetClassTopper(students);
            double average = GetClassAverage(students);
            double passPct = GetPassPercentage(students);

            Console.WriteLine("--- Class Summary ---");
            if(topper == null)
            {
                Console.WriteLine("NO students to summarize");
                return;
            }

            Console.WriteLine($"Class Topper: {topper.name} (Roll No: {topper.rollNo}, Total Marks: {topper.GetTotalMarks()})");
            Console.WriteLine($"Class Average: {average:F2}");
            Console.WriteLine($"Pass Percentage: {passPct:F2}%");
        }
    }
}
